### ISSUE TITLE
Interface extension error(by specifying some existed type)

### ISSUE THREAD

**Issue by senyaak** [Microsoft/TypeScript#14671]
**TypeScript Version:**  2.2.1 

**Code**

```ts
// this does not work
interface a {
  foo(name: string): void;
}

interface b extends a {
  foo(name: 'bar'): void;
}

// this works
interface a {
  foo(name: string): void;
}

interface b extends a {
  foo(name: 'bar'): void;
  foo(name: string): void;
}

```

**Expected behavior:**
" this does not work" - part should compile without errors
**Actual behavior:**
The compiler throw an error: 
> error TS2430: Interface 'b' incorrectly extends interface 'a'.
>   Types of property 'foo' are incompatible.
>     Type '(name: "bar") => void' is not assignable to type '(name: string) => void'.
>       Type '(name: "bar") => void' provides no match for the signature '(name: string): void'

**Comment by senyaak** [comment]
I have tried some more cases and got a questin: 
```ts

var b: B;

interface A {
  foo(name: string, callback: () => void);
}
interface B extends A {
  foo(name: 'John', callback: (age: number) => void); //foo type 1
  foo(name: string, callback: () => void); //foo type 2
}

b.foo("John",(a) => {});
```

in this case b.foo has type "foo type 2" I think the compiler should throw an error that the function ()=>void is not assignable to type (age: number) => void, shouldn't it?

**Comment by sandersn** [comment]
@senyaak I can't repro your example [on the playground](http://www.typescriptlang.org/play/index.html#src=class%20C%20%7B%20c%20%7D%0D%0Aclass%20D%20extends%20C%20%7B%20d%20%7D%0D%0Ainterface%20A%20%7B%0D%0A%20%20%20%20f(name%3A%20string)%3A%20void%3B%0D%0A%7D%0D%0Ainterface%20B%20extends%20A%20%7B%0D%0A%20%20%20%20f(name%3A%20'boo')%3A%20void%3B%0D%0A%7D#src=interface%20A%20%7B%0D%0A%20%20%20%20f(name%3A%20string)%3A%20void%3B%0D%0A%7D%0D%0Ainterface%20B%20extends%20A%20%7B%0D%0A%20%20%20%20f(name%3A%20'boo')%3A%20void%3B%0D%0A%7D)

Is there some combination of tsconfig options that I am missing?

Regarding your second question, a function with zero parameters is assignable to a function with one parameter because the zero-param function with just ignore the one argument:

```ts
type Nullary = () => void;
type Unary = (a: string) => void;
let n: Nullary;
let u: Unary;
/// not ok ///
n = u // not ok
n() // u expected an argument!
/// ok ///
u = n
u(12) // ok, n will ignore its argument
```

This is important to make `map` usable, since it passes 3 arguments to its callback, but the callback is almost always written with only 1 parameter.

**Comment by senyaak** [comment]
@sandersn Thank you for your answer.
Have only atom settings:
  "compileOnSave": false,
  "buildOnSave": false,
  "noEmitOnError": false,

in my tsconfig

> This is important to make map usable, since it passes 3 arguments to its callback, but the callback is almost always written with only 1 parameter.
I think it's a broken behavior, the code isn't type save now.. The optional parameters should be used  in this case

**Comment by senyaak** [comment]
Tried on the other pc - it works now Oo
Can be closed.

**Comment by sandersn** [comment]
Thanks for checking.

Here's the FAQ entry on why functions with fewer parmaeters are assignable to functions with more: https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-with-fewer-parameters-assignable-to-functions-that-take-more-parameters
